### PR TITLE
Firebase auth backend session cookie related set up

### PR DESCRIPTION
**Pull request by SibingWu**
Set up the `createSessionCookie`, and `clearSessionCookie` at the backend

### CODE REVIEW

File: walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+     * @param request Log in request.
+     * @return Response with session cookie.
+     */
+    @POST
+    @Path("/business-log-in")
+    @Consumes("application/json")
+    public Response createSessionCookie(HttpServletRequest request) {
+        try {
+            // Gets the ID token sent by the client
+            String idToken = getIDToken(request);
```
Comment (riya-n) [inline walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java]: nit: maybe call this function parseIDToken? just to be consistent with the other servlets

File: walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+            // Sets session expiration to 5 days
+            long expiresIn = TimeUnit.DAYS.toMillis(/* duration= */5);
```
Comment (riya-n) [inline walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java]: nit: extract the duration out into a constant?

File: walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            SessionCookieOptions options = SessionCookieOptions.builder().setExpiresIn(expiresIn).build();
+
+            // Creates the session cookie. This will also verify the ID token in the process.
+            // The session cookie will have the same claims as the ID token.
+            String sessionCookie = FirebaseAuth.getInstance().createSessionCookieAsync(idToken, options).get();
+
+            // Sets cookie policy parameters as required
+            NewCookie cookie = new NewCookie(/* name= */ "session", sessionCookie);
```
Comment (riya-n) [inline walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java]: nit: constant for "session" name?

File: walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    /**
+     * Clears the session cookie when the user signs out at the client side.
+     *
+     * @param cookie Session cookie
+     * @return Response to direct to log in page.
+     */
+    @POST
+    @Path("/business-log-out")
+    public Response clearSessionCookie(@CookieParam("session") Cookie cookie) {
+        final int maxAge = 0;
```
Comment (riya-n) [inline walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java]: what is maxAge?

File: walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        NewCookie newCookie = new NewCookie(cookie, /* comment= */ null, maxAge, /* secure= */ true);
+        return Response.temporaryRedirect(URI.create(LOG_IN_PAGE_PATH)).cookie(newCookie).build();
+    }
+
+    /** Initializes the admin SDK. */
+    private void initAdminSDK() throws IOException {
+        FirebaseOptions options = new FirebaseOptions.Builder()
+                .setCredentials(GoogleCredentials.getApplicationDefault())
+                .setDatabaseUrl("https://com-walk-in-interview.firebaseio.com/")
```
Comment (riya-n) [inline walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java]: maybe extract these out into constants as well? DATABASE_URL and PROJECT_ID

File: walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+                .setProjectId("com-walk-in-interview")
+                .build();
+
+        FirebaseApp.initializeApp(options);
+    }
+
+    /** Gets the id token from the log in request. */
+    private String getIDToken(HttpServletRequest request) throws IllegalArgumentException {
+        String idToken = request.getParameter(ID_TOKEN_PARAM).trim();
```
Comment (riya-n) [inline walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java]: maybe you can use something like: 
`        String jobIdStr = ServletUtils.getStringParameter(request, JOB_ID_FIELD, /* defaultValue= */ "");
`

File: walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        if (idToken == null || idToken.isEmpty()) {
+            throw new IllegalArgumentException("ID Token should be an non-empty string");
+        }
+
+        return idToken;
+    }
+
+    /** Gets the uid the of the current account. */
+    private Response getUid(FirebaseToken decodedToken) {
+        String uid = decodedToken.getUid();
```
Comment (riya-n) [inline walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java]: could this be null or empty?

File: walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+import javax.ws.rs.core.Cookie;
+import javax.ws.rs.core.NewCookie;
+import javax.ws.rs.core.Response;
+import java.io.IOException;
+import java.net.URI;
+import java.util.concurrent.TimeUnit;
+
+/** Util methods related to firebase auth session cookies. */
+public final class FirebaseAuthSessionCookieUtil {
+    private static final int SESSION_COOKIE_DURATION_DAY = 5;
```
Comment (riya-n) [inline walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java]: nit: SESSION_COOKIE_DURATION_DAYS

File: walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+     */
+    @POST
+    @Path("/profile")
+    public Response verifySessionCookie(@CookieParam(SESSION_COOKIE_NAME) Cookie cookie) {
+        String sessionCookie = cookie.getValue();
+        try {
+            // Verify the session cookie. In this case an additional check is added to detect
+            // if the user's Firebase session was revoked, user deleted/disabled, etc.
+            final boolean checkRevoked = true;
+            FirebaseToken decodedToken = FirebaseAuth.getInstance().verifySessionCookie(sessionCookie, checkRevoked);
```
Comment (riya-n) [inline walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java]: nit: `FirebaseAuth.getInstance().verifySessionCookie(sessionCookie, /* checkRevoked= */ true);`

**Review by riya-n — APPROVED** [review]

File: walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+     */
+    @POST
+    @Path("/business-log-in")
+    @Consumes("application/json")
+    public Response createSessionCookie(HttpServletRequest request) {
+        try {
+            // Gets the ID token sent by the client
+            String idToken = parseIDToken(request);
+
+            // Sets session expiration to 5 days
```
Comment (yohanlaunay) [inline walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java]: nit, you should remove the "5 days" from your comment., it's defined by your constant. You can just say "set the expiration".

File: walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+            // Creates the session cookie. This will also verify the ID token in the process.
+            // The session cookie will have the same claims as the ID token.
+            String sessionCookie = FirebaseAuth.getInstance().createSessionCookieAsync(idToken, options).get();
+
+            // Sets cookie policy parameters as required
+            NewCookie cookie = new NewCookie(/* name= */ SESSION_COOKIE_NAME, sessionCookie);
+            return Response.ok().cookie(cookie).build();
+        } catch (Exception e) {
+            return Response.status(Response.Status.UNAUTHORIZED).entity("Failed to create a session cookie").build();
```
Comment (yohanlaunay) [inline walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java]: please log the exception

File: walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+    public Response verifySessionCookie(@CookieParam(SESSION_COOKIE_NAME) Cookie cookie) {
+        String sessionCookie = cookie.getValue();
+        try {
+            // Verify the session cookie. In this case an additional check is added to detect
+            // if the user's Firebase session was revoked, user deleted/disabled, etc.
+            FirebaseToken decodedToken = FirebaseAuth.getInstance()
+                    .verifySessionCookie(sessionCookie, /* checkRevoked= */ true);
+            return getUid(decodedToken);
+        } catch (FirebaseAuthException e) {
+            // Session cookie is unavailable, invalid or revoked. Force user to login.
```
Comment (yohanlaunay) [inline walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java]: please log the exception

**Review by yohanlaunay — CHANGES_REQUESTED** [review]

File: walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+import javax.ws.rs.Path;
+import javax.ws.rs.core.Cookie;
+import javax.ws.rs.core.NewCookie;
+import javax.ws.rs.core.Response;
+import java.io.IOException;
+import java.net.URI;
+import java.util.concurrent.TimeUnit;
+
+/** Util methods related to firebase auth session cookies. */
+public final class FirebaseAuthSessionCookieUtil {
```
Comment (yohanlaunay) [inline walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java]: please rename to FirebaseAuthSessionCookieServlet

File: walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+     * Clears the session cookie when the user signs out at the client side.
+     *
+     * @param cookie Session cookie
+     * @return Response to direct to log in page.
+     */
+    @POST
+    @Path("/business-log-out")
+    public Response clearSessionCookie(@CookieParam(SESSION_COOKIE_NAME) Cookie cookie) {
+        // The maximum age of the cookie in seconds when the cookie will expire
+        final int maxAge = 0;
```
Comment (yohanlaunay) [inline walk-in-interview/src/main/java/com/google/account/FirebaseAuthSessionCookieUtil.java]: you need to actually clear the firebase session here:
- extract the sessionToken
- clear it with firebase
- return an expired cookie

File: walk-in-interview/src/main/java/com/google/account/CreateSessionCookieServlet.java [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+@WebServlet("/business-log-in")
+public final class CreateSessionCookieServlet extends HttpServlet {
+    private static final Logger LOGGER = Logger.getLogger(CreateSessionCookieServlet.class.getName());
+
+    private static final int SESSION_COOKIE_DURATION_DAYS = 5;
+    private static final String SESSION_COOKIE_NAME = "session";
+    private static final String ID_TOKEN_PARAM = "idToken";
+    private static final String LOG_IN_PAGE_PATH = "/log-in/index.html";
+
+    private static final String DATABASE_URL = "https://com-walk-in-interview.firebaseio.com/";
```
Comment (yohanlaunay) [inline walk-in-interview/src/main/java/com/google/account/CreateSessionCookieServlet.java:29]: `TODO(issue/) move to config file`

File: walk-in-interview/src/main/java/com/google/account/CreateSessionCookieServlet.java [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+        } catch (Exception e) {
+            String errorMessage = "Failed to create a session cookie";
+            LOGGER.log(Level.SEVERE, errorMessage, e);
+            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+            response.sendRedirect(LOG_IN_PAGE_PATH);
+        }
+    }
+
+    /** Initializes the admin SDK. */
+    private void initAdminSDK() throws IOException {
```
Comment (yohanlaunay) [inline walk-in-interview/src/main/java/com/google/account/CreateSessionCookieServlet.java]: nit: ideally this should be moved to your `FirebaseAuthUtils#initAdminSdk(databaseUrl, projectId, projectName);`

**Review by yohanlaunay — APPROVED** [review]

**Pull request merged**